core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerChannelManager.scala

```diff
@@ -98,55 +99,91 @@ object TransactionMarkerChannelManager {
+      error match {
+        case Errors.NONE =>
+          trace(s"Competed sending transaction markers for $transactionalId as $txnResult")
+
+          txnStateManager.getTransactionState(transactionalId) match {
+            case Some(epochAndMetadata) =>
+              epochAndMetadata synchronized {
+                if (epochAndMetadata.coordinatorEpoch == coordinatorEpoch) {
+                  debug(s"Updating $transactionalId's transaction state to $txnMetadata with coordinator epoch $coordinatorEpoch for $transactionalId succeeded")
+
+                  // try to append to the transaction log
+                  def retryAppendCallback(error: Errors): Unit =
+                    error match {
+                      case Errors.NONE =>
+                        trace(s"Completed transaction for $transactionalId with coordinator epoch $coordinatorEpoch, final state: state after commit: ${txnMetadata.state}")
+
+                      case Errors.NOT_COORDINATOR =>
+                        info(s"No longer the coordinator for transactionalId: $transactionalId while trying to append to transaction log, skip writing to transaction log")
+
+                      case Errors.COORDINATOR_NOT_AVAILABLE =>
+                        warn(s"Failed updating transaction state for $transactionalId when appending to transaction log due to ${error.exceptionName}. retrying")
+
+                        // retry appending
+                        txnStateManager.appendTransactionToLog(transactionalId, coordinatorEpoch, txnMetadata, retryAppendCallback)
+
+                      case errors: Errors =>
+                        throw new IllegalStateException(s"Unexpected error ${errors.exceptionName} while appending to transaction log for $transactionalId")
+                    }
+
+                  txnStateManager.appendTransactionToLog(transactionalId, coordinatorEpoch, txnMetadata, retryAppendCallback)
+                } else {
+                  info(s"Updating $transactionalId's transaction state to $txnMetadata with coordinator epoch $coordinatorEpoch for $transactionalId failed after the transaction markers " +
+                    s"has been sent to brokers. The cached metadata have been changed to $epochAndMetadata since preparing to send markers")
+                }
+              }
+
+            case None =>
+              // this transactional id no longer exists, maybe the corresponding partition has already been migrated out.
+              // return NOT_COORDINATOR to let the client re-discover the transaction coordinator
+              info(s"Updating $transactionalId's transaction state to $txnMetadata with coordinator epoch $coordinatorEpoch for $transactionalId failed after the transaction message " +
+                s"has been appended to the log. The partition ${txnStateManager.partitionFor(transactionalId)} may have migrated as the metadata is no longer in the cache")
+          }
+
+        case errors: Errors =>
```

Comment:
There's no need to have the `Errors` type parameter here. That check always succeeds.